src/packet/tcp.rs.in

```diff
@@ -0,0 +1,269 @@
+             length: vec![],
+             data: vec![],
+        }
+    }
+
+    /// Create Timestamp option
+    pub fn timestamp(my: [u8;4], their: [u8;4]) -> Self {
+        let mut data = vec![];
+        data.extend_from_slice(&my[..]);
+        data.extend_from_slice(&their[..]);
+
+        TcpOption {
+            number: TcpOptionNumbers::TIMESTAMPS,
+            length: vec![10],
+            data: data
+        }
+    }
+}
+
+/// This function gets the 'length' of the length field of the IPv4Option packet
+/// Few options (EOL, NOP) are 1 bytes long, and then have a length field equal
+/// to 0
+fn tcp_option_length(option: &TcpOptionPacket) -> usize {
+    match option.get_number() {
+        TcpOptionNumbers::EOL => 0,
+        TcpOptionNumbers::NOP => 0,
+        _ => 1,
+    }
+}
+
+fn tcp_option_payload_length(ipv4_option: &TcpOptionPacket) -> usize {
+    match ipv4_option.get_length().first() {
+        Some(len) => *len as usize - 2,
+        None => 0,
+    }
+}
+
+impl TcpOptionNumber {
+    /// Create a new TcpOptionNumber
+    pub fn new(value: u8) -> TcpOptionNumber {
+        TcpOptionNumber(value)
+    }
+}
+
+impl PrimitiveValues for TcpOptionNumber {
+    type T = (u8,);
+    fn to_primitive_values(&self) -> (u8,) {
+        (self.0,)
+    }
+}
+
+fn tcp_options_length(tcp: &TcpPacket) -> usize {
+  tcp.get_data_offset() as usize * 4 - 20
+}
+
+
+/// Calculate the checksum for a packet built on IPv4
+pub fn ipv4_checksum(packet: &TcpPacket,
+                     ipv4_source: Ipv4Addr,
+                     ipv4_destination: Ipv4Addr,
+                     next_level_protocol: IpNextHeaderProtocol)
+    -> u16be {
+    let IpNextHeaderProtocol(next_level_protocol) = next_level_protocol;
+    let mut sum = 0u32;
+
+    // Checksum pseudo-header
+    // IPv4 source
+    let octets = ipv4_source.octets();
+    sum += (octets[0] as u32) << 8 | octets[1] as u32;
+    sum += (octets[2] as u32) << 8 | octets[3] as u32;
+
+    // IPv4 destination
+    let octets = ipv4_destination.octets();
+    sum += (octets[0] as u32) << 8 | octets[1] as u32;
+    sum += (octets[2] as u32) << 8 | octets[3] as u32;
+
+    // IPv4 Next level protocol
+    sum += next_level_protocol as u32;
+
+    // TCP Length
+    sum += packet.packet().len() as u32;
+
+    // Checksum TCP header/packet
+    let mut i = 0;
+    let len = packet.packet().len();
+    while i < len && i + 1 < len {
+        sum += (packet.packet()[i] as u32) << 8 | packet.packet()[i + 1] as u32;
+        i += 2;
+    }
+    // If the length is odd, make sure to checksum the final byte
+    if len & 1 != 0 {
+        sum += (packet.packet()[len - 1] as u32) << 8;
+    }
+
+    while sum >> 16 != 0 {
+        sum = (sum >> 16) + (sum & 0xFFFF);
+    }
+
+    !sum as u16
+}
+
+#[test]
+fn tcp_header_ipv4_test() {
+    use pnet::packet::ip::IpNextHeaderProtocols;
+    use pnet::packet::ipv4::MutableIpv4Packet;
+
+    let mut packet = [0u8; 20 + 32 + 4];
```

Comment:
I don't like the use of magic numbers (here, and below). Could you change these into constants (or even variables) please?

See eg: https://github.com/libpnet/libpnet/blob/d1372bd40c654d582f62855471f51715bd2a6a47/src/test.rs#L28-L31